Ensure we don't send users to the Customiser page if the current theme supports blocks

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

In 5.9, when a block theme is active, access to the Customizer is hidden by default. There are however a few remaining links to the Customizer throughout the wp-admin interface. In particular on the Themes page, and on the Dashboard.

This PRs aims to change these links depending on the current theme. If the current theme is block-based, we send users to the `Appearance -> Editor page` instead.

Trac ticket: [https://core.trac.wordpress.org/ticket/54460](https://core.trac.wordpress.org/ticket/54460)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Implement WP_Theme::is_block_based method.
We need it to check if the theme supports block templates.

anton-vlasenko committed:
Hide `Customize` button on the /wp-admin/themes.php page if a theme is block-based.

anton-vlasenko committed:
Redirect a user to the site editor instead of hiding the customize button (for block-based themes).

anton-vlasenko committed:
1. Fix permissions.
2. Don't show Customizer on the welcome dashboard.

anton-vlasenko committed:
1. Fix permissions.
2. Don't show links to Customizer on the Theme Editor page if the current theme is block-based.

noisysocks commented on src/wp-admin/includes/dashboard.php:
```diff
@@ -1989,19 +1989,29 @@ function wp_dashboard_empty() {}
  * @since 3.3.0
```
Let's add a `@since 5.9.0` here which describes how the functionality changed.

anton-vlasenko replied:
Fixed in 2e0b4a1cfcc6a9139e39d5b1cf56438bc3451864

noisysocks commented on src/wp-includes/class-wp-theme.php:
```diff
@@ -1460,6 +1460,30 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
+	 *
+	 * @return bool
+	 */
+	public function is_block_based() {
```
How does this differ to `wp_is_block_template_theme()`? Should we maybe delete `wp_is_block_template_theme()` and instead use `wp_get_theme()->is_block_based()`?

noisysocks replied:
We should also ideally add a unit test for this new method.

anton-vlasenko replied:
> How does this differ to `wp_is_block_template_theme()`?

`wp_is_block_template_theme` allows checking the current theme only.
`WP_Theme::is_block_based` works for any instance of the `WP_Theme` class.

> Should we maybe delete wp_is_block_template_theme() and instead use wp_get_theme()->is_block_based()?

I still see value in this function. It's a handy alternative to calling `wp_get_theme()->is_block_based()` directly.
Could we possibly keep it? I agree that we don't have to duplicate the implementation, so now it uses `wp_get_theme()->is_block_based()` internally.
Would you please let me know what you think of this?

> We should also ideally add a unit test for this new method.

I 100% agree, and I encourage writing unit tests. Yesterday I had to leave early, and I didn't have time for it.
Fixed in 00cda176deaedf0642f0a4be53dbcd9d8a29df7f

noisysocks commented on src/wp-admin/theme-editor.php:
```diff
@@ -196,15 +196,28 @@
 			printf(
 				/* translators: %s: Link to Custom CSS section in the Customizer. */
 				__( 'There&#8217;s no need to change your CSS here &mdash; you can edit and live preview CSS changes in the <a href="%s">built-in CSS editor</a>.' ),
-				esc_url( add_query_arg( 'autofocus[section]', 'custom_css', admin_url( 'customize.php' ) ) )
+				$customize_url
```
I'm not sure about this one. There's no way to edit CSS in the Site Editor. Perhaps if the current theme is a block theme it would be better to hide this notice altogether?

anton-vlasenko replied:
Good catch. You are right. Fixed in 58f440f5a0111fa0b523154fae65b2e5732a4748

noisysocks reviewed (commented):
Nice work. I left some notes.

anton-vlasenko committed:
Better PHPDoc block description.

anton-vlasenko committed:
Don't display the hint for block-based themes.

anton-vlasenko committed:
Don't duplicate code.
Let's use WP_Theme::is_block_based inside the wp_is_block_template_theme function to determine whether a theme is a block-based theme or not.

anton-vlasenko committed:
Add a unit test for WP_Theme::is_block_based method.

anton-vlasenko committed:
Merge branch 'trunk' into fix/hide-customize-links-for-non-block-based-themes

anton-vlasenko committed:
Fix the failing unit tests.

anton-vlasenko commented:
I've added 2 new test block themes. I tried to modify the old existing test themes and this caused many unit tests to fail. They are not intended to be the block themes (despite their names).

noisysocks commented on src/wp-includes/theme.php:
```diff
@@ -4084,6 +4084,5 @@ function create_initial_theme_features() {
  */
 function wp_is_block_template_theme() {
-	return is_readable( get_theme_file_path( '/block-templates/index.html' ) ) ||
-		is_readable( get_theme_file_path( '/templates/index.html' ) );
```
`is_block_based()` doesn't have this check for `/templates/index.html`, so not sure if we should use `wp_get_theme()→is_block_based()` as the implementation for `wp_is_block_template_theme()`. (I don't know what `/templates/index.html` is.)

noisysocks replied:
I've reverted this change for now. Let's look at simplifying in a follow-up after beta 1.

anton-vlasenko replied:
@noisysocks 

>(I don't know what /templates/index.html is.)

I think that's because we use `/templates/` instead of `/block-templates` folder to store templates now.
More info can be found here: https://github.com/WordPress/wordpress-develop/commit/b90fc1b6dda3bfff668ad07e25da545c3eb10423
I've updated the `WP_Theme::is_block_based` method, so now it works the same way as the current implementation of `wp_is_block_template_theme`.
I think it's not critical for WordPress 5.9 whether we use `wp_get_theme()→is_block_based()` internally or the current implementation (checking files with `is_readable` function). `wp_is_block_template_theme`'s internal implementation can be anything we want. We will not break backward compatibility if we change that implementation later.
I'd like to open a new trac issue to get more feedback on changing the implementation of `wp_is_block_template_theme` and `get_theme_file_path` (because this PR introduces new `WP_Theme::get_file_path` method as well).
In my opinion this PR can be merged now.

noisysocks committed:
Don't use wp_get_theme() in wp_is_block_template_theme()

anton-vlasenko committed:
Merge the changes from trunk.

anton-vlasenko committed:
1. Add a new method WP_Theme::get_file_path.
2. Check for /templates/index.html file inside the WP_Theme::is_block_based method.

anton-vlasenko committed:
Add unit tests for WP_Theme::get_file_path method.

anton-vlasenko committed:
Rename block-templates to templates.
See https://github.com/WordPress/wordpress-develop/commit/b90fc1b6dda3bfff668ad07e25da545c3eb10423 for more details.

palmsout commented:
Sorry if I'm missing the point here, but aren't there still important applications of the customizer for block themes? The obviously one being that it allows users to add additional css without having to edit theme code? I don't know where I'd be without that ability, and I use a block theme.

hellofromtonya commented on src/wp-includes/class-wp-theme.php:
```diff
@@ -1460,6 +1460,64 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
+		);
+
+		foreach ( $paths_to_index_block_template as $path_to_index_block_template ) {
+			if ( is_file( $path_to_index_block_template ) && is_readable( $path_to_index_block_template ) ) {
```
`is_file()` also checks (like `is_readable()`) that the file name given exists AND that it's a "regular file." As this is a HTML file, the check of "regular file" seems unnecessary. Though not critical, it could confuse contributors as to why it's needed.

@anton-vlasenko do you foresee the possibility of the HTML file not being a regular file?

anton-vlasenko replied:
@hellofromtonya 
First of all, thank you for the review.
The reason to use `is_file` is that someone might create a directory named `index.html`.
Of course, this is unlikely to happen.
But, if we are expecting a file, I thought it logical to check that we are dealing with a file (not a directory).
Please feel free to remove the `is_file( $path_to_index_block_template ) &&` check if you think it's redundant (or please let me know if you want me to do it).

hellofromtonya replied:
Huh, you're right. And even though there's a low chance, it's not impossible or completely unlikely it could happen. Thanks for pointing that out. Yes, let's keep it in just in case to guard against that possibility.

hellofromtonya commented on src/wp-includes/class-wp-theme.php:
```diff
@@ -1460,6 +1460,64 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
+		$stylesheet_directory = $this->get_stylesheet_directory();
+		$template_directory   = $this->get_template_directory();
+
+		if ( empty( $file ) ) {
+			$path = $stylesheet_directory;
+		} elseif ( file_exists( $stylesheet_directory . '/' . $file ) ) {
+			$path = $stylesheet_directory . '/' . $file;
+		} else {
+			$path = $template_directory . '/' . $file;
+		}
```
What if the theme or requested file does not exist?

For example: What happens if the theme doesn't exist? What happens if the file does not exist in the child or parent? In both cases, it flows through to the `else` and returns a path to a non-existent file.

Should this method check if the file exists before returning the path to it? There are checks in the `is_block_based()` method. But this method is public and could be invoked in a plugin, theme, or later in Core.

I wonder: Should it check in this method to avoid pushing the exists/readable checks downstream?

```php

// If the file does not exist, bail out and return an empty string.
if ( ! is_readable( $path ) ) {
	return '';
}
```

anton-vlasenko replied:
@hellofromtonya 

> Should this method check if the file exists before returning the path to it?

The code of this method was copied over from the [`get_theme_file_path`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/link-template.php#L4485) function.
`WP_Theme::get_file_path` method is just an object-oriented version of that function (just as the `WP_Theme::is_block_based` method is an OO version of the [`wp_is_block_template_theme`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/theme.php#L4093) function).
The reason to introduce both `WP_Theme::get_file_path` and `WP_Theme::is_block_based` methods was that we have to be able to check if any given theme supports blocks (not just the current theme) on the `/wp-admin/themes.php` page.
In my opinion, these OO methods have to behave the same as their procedural counterparts (for backward compatibility).
If/when this PR gets merged into Core, I would like to create another trac issue to refactor  `get_theme_file_path` and `wp_is_block_template_theme` functions and make them call `WP_Theme::get_file_path` and  `WP_Theme::is_block_based` internally (to avoid code duplication).

> Should it check in this method to avoid pushing the exists/readable checks downstream?

We can add a second parameter to the `WP_Theme::get_file_path` method (equals false by default) if that is what we want. If that second parameter is true, we can return null if the file doesn't exist. Or ideally, we could add another public method to the `WP_Theme` class to implement this logic (to keep the number of arguments [as low as possible](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally)).

Would you please let me know what your opinion on this matter is? Thank you.

hellofromtonya replied:
Thanks @anton-vlasenko for thoroughly explaining the reasoning for the code. Yes, I agree it makes sense (a) for these methods to be here in this class and (b) that this method uses a similar expected pattern of `get_theme_file_path()`. 

It could be interesting to add the option or another method that checks if the files exist. But I think for this PR what you have works, has solid reasons for them to be added, and fits what is needed for the attached issue. No need to add more than that at this point.

hellofromtonya committed:
Uses a data providers, adds covers, add datasets.

hellofromtonya commented:
@anton-vlasenko I refactored the tests to use data providers and named datasets as well as added additional test data for when a theme or file does not exist. Tests pass https://github.com/WordPress/wordpress-develop/pull/1933/commits/641751b5915da26eb3f014ad8f776442ba726cde.

The failed CI jobs are not related to this PR, but rather are due to changeset https://core.trac.wordpress.org/changeset/52272.

hellofromtonya reviewed (approved):
Good work @anton-vlasenko! Ready for `commit` 👍

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/52279.

anton-vlasenko commented:
Thank you, @hellofromtonya!
